### PR TITLE
Allow clustering updates in AnnData file upload (SCP-5039)

### PR DESCRIPTION
**Pull request by ehanna4**
Allow multiple clusterings, updating and deleting clusterings for an AnnData file upload

To test:
- Pull this branch and start up delayed jobs and a rails server
- Ensure you have the flag for AnnData ingest flipped to true
- Start from a fresh study (necessary as there was a bug in the way clusterings were handled previously)
- Fill in expression and clustering info
- For clustering start with one and make the name "first" and the obsm key "X_umap"
- upload the AnnData file 
- allow this to complete uploading and parsing

To test updating a name
- Now go to the clustering page and change the name of the clustering and click save
- Observe in `development.log` that an ingest job will be kicked off to re-parse this
- Observe in the Study details page the clustering will be deleted and then replaced with the new name

To test adding another clustering
- go to the clustering page and click to add a clustering
- name it "second" and give it an obsm key name of "X_tsne" 
- click save
- now wait, locally this takes a bit of time for me, but eventually you will see the job in `development.log` and refreshing the study details page a few times you will see this new clustering option available.

You can also delete a clustering (it won't present a delete button if there is only one though) and you will see this reflected in the study details page as well.

Here is a screenrecording of me going through all this - warning it's 16min haha this is not quick on local https://drive.google.com/file/d/1egNdyYy081KW5xHPdlsBAsFn3XWOOzUk/view?usp=sharing

### CODE REVIEW

File: app/models/ann_data_file_info.rb
Comment (bistline) [inline app/models/ann_data_file_info.rb:112]: I had forgotten I sent this code to you... I included it on my draft PR that covers expression parsing (due to the same bug we found).  Depending on who merges first, we'll need to remove or deal with the conflict then.

File: app/models/study_file.rb
Comment (bistline) [inline app/models/study_file.rb:1406]: This looks like a copy of `AnnDataIngestParameters#fragment_file_gs_url` - we should reference that method rather than copying it here.  That being said, if we do end up going the above route, then we wouldn't need this here.

**Review by bistline — CHANGES_REQUESTED** [review]
See the note in `study_file.rb` about renaming database objects rather than the full delete/re-parse job.

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (eweitz) [inline app/javascript/components/upload/ExpandableFileForm.jsx:228]: Blocker: the current code makes `alreadyUploaded` always `undefined`, and thus always falsy, which I believe makes `getIsUploadEnabled` always return true, which would be a bug.

For reference:
```
allFiles = [  
  {name: 'foo', file_type: 'Expression Matrix'},
  {name: 'bar', file_type: 'AnnData', ann_data_file_info: {has_clusters: false}},
  {name: 'baz', file_type: 'Expression Matrix'},
  {name: 'moo', file_type: 'Metadata'},
  {name: 'quix', file_type: 'AnnData', ann_data_file_info: {has_clusters: true}}
]

alreadyUploaded = allFiles.forEach(fileObj => {
  if (fileObj?.ann_data_file_info?.has_clusters) {
    return true
  }
})

alreadyUploaded
undefined
```

I think you intend `some` or `all` instead of `forEach`, and a different inner function.  For reference:

```
// Some AnnData files are uploaded (assuming `has_clusters: true` reliably indicates file is uploaded)
allFiles = [  
  {name: 'foo', file_type: 'Expression Matrix'},
  {name: 'bar', file_type: 'AnnData', ann_data_file_info: {has_clusters: false}},
  {name: 'baz', file_type: 'Expression Matrix'},
  {name: 'moo', file_type: 'Metadata'},
  {name: 'quix', file_type: 'AnnData', ann_data_file_info: {has_clusters: true}}
]

alreadyUploaded = allFiles.some(fileObj => fileObj?.ann_data_file_info?.has_clusters)

alreadyUploaded
true
```

I don't have a complete enough understanding to confidently assert which of `some` or `all` is appropriate, but I do think there's a bug (or at least unintentionally misleading code) here.

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (eweitz) [inline app/javascript/components/upload/ExpandableFileForm.jsx:252]: More concisely, and idiomatically:

```suggestion
  // only allow save/delete for updates after an AnnData file has been uploaded
  const alreadyUploaded = 
    allFiles.some(fileObj => fileObj?.ann_data_file_info?.data_fragments.length > 0)
```

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (eweitz) [inline app/javascript/components/upload/ExpandableFileForm.jsx:266]: Less cyclomatic complexity, more concision:

```suggestion
  if ((isClustering || isExpressionMatrix || isAnnData) && alreadyUploaded && isAnnDataExperience) {
    return true
  } else {
    return isAnnData && isAnnDataExperience
  }
```

File: app/javascript/components/upload/UploadWizard.jsx
Comment (eweitz) [inline app/javascript/components/upload/UploadWizard.jsx:92]: For maintainability, let's either use:

```suggestion
  const [isOverrideExperienceMode, setIsOverrideExperienceMode] = useState(false)
```

like `const [isAnnDataExperience, setIsAnnDataExperience] = useState(false)`, or let's make the latter `const [annDataExperience, setAnnDataExperience] = useState(false)`.  

I personally think `isFoo` for ideal for booleans, but I don't feel strongly.  My only qualm here is with the locally inconsistent naming style.

File: app/javascript/components/upload/UploadWizard.jsx
Comment (eweitz) [inline app/javascript/components/upload/UploadWizard.jsx:280]: Could you break this up with some punctuation?  

The comment strikes me as important, but I've tried reading it a few times, and I unfortunately can't really understand it.  (I don't intend to wordsmith code comments; I'm just seeking to ensure maintainers can clearly understand what's written here.)

File: app/javascript/components/upload/UploadWizard.jsx
Comment (eweitz) [inline app/javascript/components/upload/UploadWizard.jsx:301]:
```suggestion
              // multiple clustering forms are allowed so add each as a fragment to the AnnData file
```

File: app/javascript/components/upload/UploadWizard.jsx
Comment (eweitz) [inline app/javascript/components/upload/UploadWizard.jsx:277]: For maintainability, could you extract lines 278 - 314 into a separate function?

Those 36 lines add substantially complexity to this `saveFile` function that's already likely a bit too complex.  Having the `isAnnDataExperience` logic in its own function would enable the containing `saveFile` function to be understand more easily at a high level, while also making it more testable (which would help robustness).

**Review by eweitz — COMMENTED** [review]
Good progress!

I got about halfway through the review before you conscientiously noted it's worth ceasing review, as a chat with Jon entails substantial changes here.  I suspect my findings so far are mostly orthogonal to that, so I've noted them here.  

I found two trivial blockers that seem like they'd cause user-facing bugs.  I also suggest a handful of non-blocking refinements for maintainability.  All specific code suggestions are just for convenience, and worth a quick check if accepted.

File: app/javascript/components/upload/UploadWizard.jsx
Comment (ehanna4) [reply app/javascript/components/upload/UploadWizard.jsx:92]: Oh in my mind I was reading 'override' as the action verb like override the experience mode? Yes or no. But I see how that could be confusing since 'override' could be read as a noun like is the override-experience-mode on or off? I can change it.

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (ehanna4) [reply app/javascript/components/upload/ExpandableFileForm.jsx:228]: ah i think that was actually a leftover from a previous attempt that I missed removing, good catch!

File: app/models/study_file.rb
Comment (bistline) [inline app/models/study_file.rb:1338]: There are some convenience methods that will make this simpler, I think.  `StudyFile#is_reference_anndata?` also calls `is_anndata?`, and all `AnnData` files will already have an `AnnDataFileInfo` object initialized:
```suggestion
    if !!is_reference_anndata? && ann_data_file_info.data_fragments_changed?
```
That will cover the 4 checks you already had, and `!!` does the same thing in Ruby it does in Javascript.

Also (I know I used to be bad about this, and you probably copied stuff of mine from elsewhere), you don't need to call `self` in Ruby unless you're doing a direct assignment in an initializer or need to pass a reference.  Any method/attribute can be called directly  - `file_type` and `is_reference_anndata?` are the same as `self.file_type` and `self.is_reference_anndata?` - and the `self` is implicit, depending on the method scope.

File: app/models/study_file.rb
Comment (bistline) [inline app/models/study_file.rb:1358]: Blocker: this query needs to include the previous name of the `ClusterGroup` as well as the `study_file` (and here we do need `self`).  Also, we can just pass the study association to the query for the sake of brevity.  And I find `dig` easier to work with than multiple `[]` calls:
```suggestion
              @cluster = ClusterGroup.find_by(study:, study_file: self, name: prev_ids.dig(curr_id, 'name'))
```

File: app/lib/file_parse_service.rb
Comment (bistline) [inline app/lib/file_parse_service.rb:13]: Blocker: this method declaration needs `obsm_key: nil` added to it so we can call it from `StudyFile#call_ingest_on_updated_clustering_fragments`.

File: app/lib/file_parse_service.rb
Comment (bistline) [inline app/lib/file_parse_service.rb:108]:
```suggestion
          if obsm_key.present?
```

File: app/models/study_file.rb
Comment (bistline) [inline app/models/study_file.rb:1367]:
```suggestion
            FileParseService.run_parse_job(self, study, study.user, obsm_key: fragment)
```

File: app/models/study_file.rb
Comment (bistline) [inline app/models/study_file.rb:1337]: The method name seems pretty long (though descriptive).  Maybe `handle_data_fragment_updates` since some are just database calls and others new parse jobs?

**Review by bistline — APPROVED** [review]
Two blockers, but both pretty straightforward so I'm preemptively approving so you're not blocked tomorrow.  Thanks for working this all out!

File: app/models/study_file.rb
Comment (ehanna4) [reply app/models/study_file.rb:1337]: Since this method (currently) only does stuff on clustering fragments I'll go with `handle_clustering_fragment_updates`

**Review by eweitz — CHANGES_REQUESTED** [review]
Getting closer!  This is a tough area of the code, kudos on keeping with it.

I see only one significant blocker: the `deleteFile` function needs rework.  The [biggest issue is uploading content as part of deleting it](https://github.com/broadinstitute/single_cell_portal_core/pull/1771#discussion_r1186407623), which strikes me as needing non-trivial effort to resolve, and also of significant functional impact for the end user.  (Ideally, I'm wrong about one or both of those concerns.)

Also, not quite blocking, but I think we need some frontend testing for these AnnData updates.  The JavaScript is complex.  With no tests for this significant new JS, I'm concerned we won't be able to change it confidently and quickly.  We ought to be able to use mocks in Jest to mimic server responses if needed; I don't think more realistic end-to-end tests are required.

I'm free to pair any time for help with these!

Finally, thanks for the demo video.  Almost 17 minutes of video-only recording is too long and raw to be very useful for product or technical review, but it _does_ seem reasonably helpful as a full-fidelity reference for maintenance.  One idea to make such resources more valuable without much effort: consider adding timestamped links to salient parts (e.g., [Update cluster in AnnData file at 5m30s](https://drive.google.com/file/d/1egNdyYy081KW5xHPdlsBAsFn3XWOOzUk/view?t=5m30s)).  This is just food for thought, not a request for refinement!

File: app/javascript/components/upload/UploadWizard.jsx
Comment (ehanna4) [reply app/javascript/components/upload/UploadWizard.jsx:240]: This isn't my comment, I think the linter picked it up or maybe in clean up I just fixed the missing capitalization.

Comment (eweitz) [reply app/javascript/components/upload/UploadWizard.jsx:240]: Two clicks to accept this suggestion and pull strikes me as a decent effort-value ratio to address this technically-incorrect comment (which I appreciate wasn't written by you!) -- but that said, this isn't a blocker.

File: app/javascript/components/upload/UploadWizard.jsx
Comment (ehanna4) [reply app/javascript/components/upload/UploadWizard.jsx:174]: Made a ticket to more holistically address this: https://broadinstitute.slack.com/archives/CBEHTH601/p1683320088149879

Seems reasonable but would prefer to have done in a different ticket.

File: app/javascript/components/upload/UploadWizard.jsx
Comment (ehanna4) [reply app/javascript/components/upload/UploadWizard.jsx:450]: Mentioned in other comment - seems like a totally reasonable change but not something I would like to address in this PR so made a ticket to address: https://broadinstitute.slack.com/archives/CBEHTH601/p1683320088149879

File: app/javascript/components/upload/UploadWizard.jsx
Comment (ehanna4) [reply app/javascript/components/upload/UploadWizard.jsx:431]: https://broadinstitute.slack.com/archives/CBEHTH601/p1683320088149879

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (ehanna4) [reply app/javascript/components/upload/ExpandableFileForm.jsx:156]: Oh woops got lost in the mix, addressing this now

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (ehanna4) [reply app/javascript/components/upload/ExpandableFileForm.jsx:242]: Hmm that's a good thought. I'm not totally sure how this piece would work within the environment of multiple AnnData files. Using the logic as is with `every` and multiple AnnDatas would mean you couldn't save on any AnnData file if there is another AnnData file not yet uploaded but is being filled in. Whereas exact logic with `some` would allow saving on AnnDatas you are currently uploading as long as one is up. So both not quite right, with `some` being too permissive and `every` being too restrictive.

So I think instead of looking at `allFiles` when we go multi-AnnData we'd have to add some logic for each individual AnnData. 

So I think you are correct that `some` won't play right if keeping this exact logic when allowing multiple AnnData but I don't think `every` would work exactly right either and so either option will require some re-working to support multi.

**Review by eweitz — APPROVED** [review]
Thanks for investigating further!

It seems a trivial blocker might have been overlooked, see "Blocker: this saveButton assignment".

My "Request changes" blocking concern about re-uploading whole AnnData files to delete a cluster is indeed resolved.  Code comments suggested below would help maintainability, but I don't consider that blocking.

File: app/javascript/components/upload/ExpandableFileForm.jsx
Comment (ehanna4) [reply app/javascript/components/upload/ExpandableFileForm.jsx]: Hey I just updated this a bit to deal better with showing parsing status for fragments and added some comments to make it a bit more clear what the conditional saveButton assignments do and re-arranged a bit so I think it's fixed. 

Unfortunately now the lines aren't matching up so I can't 100% confirm to these lines but what is now on line 158 is the initial `saveButton` after that it is only updated when saving, parsing or saveDisabled.

Comment (eweitz) [reply app/javascript/components/upload/ExpandableFileForm.jsx]: Awesome, that looks good!

**Pull request merged**
